### PR TITLE
MKL-DNN LBR-GRU Inference Integration (FP32 LBR-GRU)

### PR DESCRIPTION
**Pull request by zixuanweeei**
## Description ##
We integrated the mkl-dnn Linear-Before-Reset GRU into MXNet. Currently, it supports FP32 inference. @pengzhao-intel @ciyongch @TaoLv 

### Essentials ###
- [x] Changes are complete. The FP32 inference of all the RNN variants supported by MXNet is ready in this PR. 

### Changes ###
- [x] LBR-GRU inference goes directly into mkl-dnn rnn forward primitive by default.
- [x] Move `mkldnn::memory`s into a struct.
- [x] Drop unnecessary `mkldnn::memory`s from multi-layer implementation.

### Performance ###
We tested the performance of FusedRNN with `mode='gru'` using the same dimension as that in PR#14713, i.e. seq_length = 300, batch_size = 20, input_size = 800, hidden_size = 800.
<table border="0">
 <tr height=19 style='height:14.4pt'>
  <td rowspan=2 height=38 class=xl95 align=center width=26>mode</td>
  <td rowspan=2 class=xl95 width=39 align=center>Layer</td>
  <td rowspan=2 class=xl95 width=61 align=center>Direction</td>
  <td colspan=2 class=xl90 width=241 align=center>MXNET_USE_MKLDNN_RNN=0</td>
  <td colspan=2 class=xl90 width=241 align=center>MXNET_USE_MKLDNN_RNN=1</td>
  <td colspan=2 class=xl90 width=172 align=center>SpeedUp</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl90 align=center>Throughput (samples/sec)</td>
  <td class=xl90 align=center>Latency (ms)</td>
  <td class=xl90 align=center>Throughput (samples/sec)</td>
  <td class=xl90 align=center>Latency (ms)</td>
  <td class=xl90 align=center>Throughtput</td>
  <td class=xl90 align=center>Latency</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl70>gru</td>
  <td class=xl70 align=right>1</td>
  <td class=xl70 align=right>1</td>
  <td class=xl71 align=right>430.03</td>
  <td class=xl72 align=right>20.43</td>
  <td class=xl71 align=right>806.27</td>
  <td class=xl72 align=right>4.28</td>
  <td class=xl78 align=right>1.87</td>
  <td class=xl78 align=right>4.78</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl70>gru</td>
  <td class=xl70 align=right>1</td>
  <td class=xl70 align=right>2</td>
  <td class=xl71 align=right>218.58</td>
  <td class=xl72 align=right>119.50</td>
  <td class=xl71 align=right>416.55</td>
  <td class=xl72 align=right>8.58</td>
  <td class=xl78 align=right>1.91</td>
  <td class=xl78 align=right>13.93</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl70>gru</td>
  <td class=xl70 align=right>5</td>
  <td class=xl70 align=right>1</td>
  <td class=xl71 align=right>89.47</td>
  <td class=xl72 align=right>100.07</td>
  <td class=xl71 align=right>177.52</td>
  <td class=xl72 align=right>21.20</td>
  <td class=xl78 align=right>1.98</td>
  <td class=xl78 align=right>4.72</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl70>gru</td>
  <td class=xl70 align=right>5</td>
  <td class=xl70 align=right>2</td>
  <td class=xl71 align=right>39.68</td>
  <td class=xl72 align=right>611.38</td>
  <td class=xl71 align=right>71.15</td>
  <td class=xl72 align=right>46.45</td>
  <td class=xl78 align=right>1.79</td>
  <td class=xl78 align=right>13.16</td>
 </tr>
</table>

We also compared the performance of this PR with that of the previously integrated LSTM, vRNN tanh, vRNN Relu on branch master. It seems that there is a distinct regression with `mode='lstm'`. 

<table border="0">
 <tr height=19 style='height:14.4pt'>
  <td rowspan=2 height=38 class=xl95 width=60 align=center>mode</td>
  <td rowspan=2 class=xl95 width=39 align=center>Layer</td>
  <td rowspan=2 class=xl95 width=61 align=center>Direction</td>
  <td colspan=2 class=xl97 width=241 align=center>eec0fb4</td>
  <td colspan=2 class=xl97 width=241 align=center>This PR (c186863)</td>
  <td colspan=2 class=xl90 width=172 align=center>Gap</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl96 align=center>Throughput (samples/sec)</td>
  <td class=xl96 align=center>Latency (ms)</td>
  <td class=xl96 align=center>Throughput (samples/sec)</td>
  <td class=xl96 align=center>Latency (ms)</td>
  <td class=xl78 align=center>Throughput</td>
  <td class=xl78 align=center>Latency</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>lstm</td>
  <td class=xl78 align=right>1</td>
  <td class=xl78 align=right>1</td>
  <td class=xl73 align=right>675.24</td>
  <td class=xl73 align=right>4.98</td>
  <td class=xl73 align=right>654.61</td>
  <td class=xl73 align=right>5.71</td>
  <td class=xl73 align=right>0.97</td>
  <td class=xl73 align=right>0.87</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>lstm</td>
  <td class=xl78 align=right>1</td>
  <td class=xl78 align=right>2</td>
  <td class=xl73 align=right>343.99</td>
  <td class=xl73 align=right>9.86</td>
  <td class=xl73 align=right>333.13</td>
  <td class=xl73 align=right>11.65</td>
  <td class=xl73 align=right>0.97</td>
  <td class=xl73 align=right>0.85</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>lstm</td>
  <td class=xl78 align=right>5</td>
  <td class=xl78 align=right>1</td>
  <td class=xl73 align=right>141.30</td>
  <td class=xl73 align=right>24.03</td>
  <td class=xl73 align=right>138.59</td>
  <td class=xl73 align=right>28.39</td>
  <td class=xl73 align=right>0.98</td>
  <td class=xl73 align=right>0.85</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>lstm</td>
  <td class=xl78 align=right>5</td>
  <td class=xl78 align=right>2</td>
  <td class=xl73 align=right>55.67</td>
  <td class=xl73 align=right>53.16</td>
  <td class=xl73 align=right>54.11</td>
  <td class=xl73 align=right>61.29</td>
  <td class=xl73 align=right>0.97</td>
  <td class=xl73 align=right>0.87</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>rnn_tanh</td>
  <td class=xl78 align=right>1</td>
  <td class=xl78 align=right>1</td>
  <td class=xl73 align=right>1617.27</td>
  <td class=xl73 align=right>2.46</td>
  <td class=xl73 align=right>1541.13</td>
  <td class=xl73 align=right>2.60</td>
  <td class=xl73 align=right>0.95</td>
  <td class=xl73 align=right>0.94</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>rnn_tanh</td>
  <td class=xl78 align=right>1</td>
  <td class=xl78 align=right>2</td>
  <td class=xl73 align=right>851.16</td>
  <td class=xl73 align=right>4.82</td>
  <td class=xl73 align=right>828.10</td>
  <td class=xl73 align=right>5.01</td>
  <td class=xl73 align=right>0.97</td>
  <td class=xl73 align=right>0.96</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>rnn_tanh</td>
  <td class=xl78 align=right>5</td>
  <td class=xl78 align=right>1</td>
  <td class=xl73 align=right>390.48</td>
  <td class=xl73 align=right>11.66</td>
  <td class=xl73 align=right>376.38</td>
  <td class=xl73 align=right>12.27</td>
  <td class=xl73 align=right>0.96</td>
  <td class=xl73 align=right>0.95</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>rnn_tanh</td>
  <td class=xl78 align=right>5</td>
  <td class=xl78 align=right>2</td>
  <td class=xl73 align=right>164.11</td>
  <td class=xl73 align=right>25.72</td>
  <td class=xl73 align=right>156.64</td>
  <td class=xl73 align=right>26.74</td>
  <td class=xl73 align=right>0.95</td>
  <td class=xl73 align=right>0.96</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>rnn_relu</td>
  <td class=xl78 align=right>1</td>
  <td class=xl78 align=right>1</td>
  <td class=xl73 align=right>1582.22</td>
  <td class=xl73 align=right>2.65</td>
  <td class=xl73 align=right>1508.54</td>
  <td class=xl73 align=right>2.59</td>
  <td class=xl73 align=right>0.95</td>
  <td class=xl73 align=right>1.02</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>rnn_relu</td>
  <td class=xl78 align=right>1</td>
  <td class=xl78 align=right>2</td>
  <td class=xl73 align=right>824.18</td>
  <td class=xl73 align=right>5.20</td>
  <td class=xl73 align=right>803.40</td>
  <td class=xl73 align=right>5.04</td>
  <td class=xl73 align=right>0.97</td>
  <td class=xl73 align=right>1.03</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>rnn_relu</td>
  <td class=xl78 align=right>5</td>
  <td class=xl78 align=right>1</td>
  <td class=xl73 align=right>381.53</td>
  <td class=xl73 align=right>12.58</td>
  <td class=xl73 align=right>366.71</td>
  <td class=xl73 align=right>12.09</td>
  <td class=xl73 align=right>0.96</td>
  <td class=xl73 align=right>1.04</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl78>rnn_relu</td>
  <td class=xl78 align=right>5</td>
  <td class=xl78 align=right>2</td>
  <td class=xl73 align=right>153.11</td>
  <td class=xl73 align=right>27.57</td>
  <td class=xl73 align=right>153.67</td>
  <td class=xl73 align=right>27.06</td>
  <td class=xl73 align=right>1.00</td>
  <td class=xl73 align=right>1.02</td>
 </tr>
</table>

## Comments ##
- The gates orders of GRU from MXNet and MKL-DNN are different. There are more overhead costs when it prepares `mkldnn::memory`s with `mode='gru'`.

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 1]
```diff
@@ -118,7 +151,7 @@ static size_t GetMKLDNNRNNCacheMemorySize(int L,
       break;
     case rnn_enum::kGru:
       size = 2 * (D * (I + H) * 3 * H + (L - 1) * D * (D * H + H) * 3 * H +
-             L * D * 2 * N * H) + T * N * D * H + L * 2 * D * 3 * H + (L + 2) * D * 2 * N * H +
+             L * D * 2 * N * H) + T * N * D * H + L * 2 * D * 4 * H + (L + 2) * D * 2 * N * H +
```
Comment (marcoabreu) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Nit: I know it's out of the scope of this PR, but could we rename the variables to something more self-explanatory?

**Comment by marcoabreu** [comment]
Could you elaborate where "MXNET_USE_MKLDNN_RNN" comes from? Out of scope of the PR, but why did we introduce that switch instead of just going with it?

**Comment by pengzhao-intel** [comment]
> Could you elaborate where "MXNET_USE_MKLDNN_RNN" comes from? Out of scope of the PR, but why did we introduce that switch instead of just going with it?

@marcoabreu some background of this env variable. We had pulled in the MKL-DNN RNN integration in 1.5 by the application team for RNN/LSTM. So we set an env to make a rollback for the user in case there are un-expected functionality and performance degree. Now, we are refactoring the code and add the whole support for RNN API. This variable will be removed in the near future.

What's your opinion?

**Comment by marcoabreu** [comment]
That approach sounds excellent and very clean, great idea! Thanks for elaborating.

Near future is before 1.6, correct?

**Comment by pengzhao-intel** [comment]
> That approach sounds excellent and very clean, great idea! Thanks for elaborating.
> 
> Near future is before 1.6, correct?

Yes, before 1.6 :)

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 1 above)
Comment (zixuanweeei) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: No problem. I will have a try. I ever did it. But it seems that there is a large amount of code. I think we can segregate the common parts from them.

Comment (zixuanweeei) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: @marcoabreu Sorry for the late update. I have renamed this part. Would you mind checking it again? Thanks.

**Comment by abhinavs95** [comment]
@mxnet-label-bot add [mkldnn, pr-work-in-progress]

**Comment by zixuanweeei** [comment]
Performance of the latest commit. @ciyongch I have checked the performance again. Their perf are similar, same as we discussed last time. 
<table border="0">
  <td rowspan=2 height=39 class=xl92 width=60 style='height:29.4pt;width:45pt' align=center>Mode</td>
  <td rowspan=2 class=xl92 width=39 style='width:29pt' align=center>Layer</td>
  <td rowspan=2 class=xl92 width=61 style='width:46pt' align=center>Direction</td>
  <td rowspan=2 class=xl92 width=106 style='width:80pt' align=center>Dims</td>
  <td colspan=2 class=xl91 width=241 style='border-left:none;width:181pt' align=center> a26af2b </td>
  <td colspan=2 class=xl91 width=241 style='border-left:none;width:181pt' align=center>This PR ( cfc6910 )</td>
  <td colspan=2 class=xl90 width=127 style='border-left:none;width:96pt' align=center>Gap</td>
 </tr>
 <tr>
  <td>Throughput (samples/sec)</td>
  <td>Latency (ms)</td>
  <td>Throughput (samples/sec)</td>
  <td>Latency (ms)</td>
  <td>Throughput</td>
  <td>Latency</td>
 </tr>
 <tr>
  <td>lstm</td>
  <td>1</td>
  <td>1</td>
  <td class=xl74>50, 128, 800, 800</td>
  <td class=xl75 >630.78</td>
  <td class=xl76 >4.82</td>
  <td class=xl75 >670.23</td>
  <td class=xl76 >4.87</td>
  <td class=xl94 >1.06</td>
  <td class=xl95 >0.99</td>
 </tr>
 <tr>
  <td height=20 class=xl77 style='height:15.0pt;border-top:none'>lstm</td>
  <td class=xl67>1</td>
  <td class=xl67>2</td>
  <td class=xl6>50, 128, 800, 800</td>
  <td class=xl68>313.71</td>
  <td class=xl69>9.68</td>
  <td class=xl68>338.51</td>
  <td class=xl69>9.72</td>
  <td class=xl96>1.08</td>
  <td class=xl95 >1.00</td>
 </tr>
 <tr>
  <td height=20 class=xl77 style='height:15.0pt;border-top:none'>lstm</td>
  <td class=xl67>5</td>
  <td class=xl67>1</td>
  <td class=xl6>50, 128, 800, 800</td>
  <td class=xl68>139.85</td>
  <td class=xl69>23.59</td>
  <td class=xl68>138.22</td>
  <td class=xl69>23.48</td>
  <td class=xl96>0.99</td>
  <td class=xl95 >1.00</td>
 </tr>
 <tr>
  <td height=20 class=xl77 style='height:15.0pt;border-top:none'>lstm</td>
  <td class=xl67>5</td>
  <td class=xl67>2</td>
  <td class=xl6>50, 128, 800, 800</td>
  <td class=xl68>54.63</td>
  <td class=xl69>51.19</td>
  <td class=xl68>54.27</td>
  <td class=xl69>51.28</td>
  <td class=xl96>0.99</td>
  <td class=xl95 >1.00</td>
 </tr>
 <tr>
  <td height=20 class=xl73 style='height:15.0pt'>rnn_tanh</td>
  <td class=xl74 >1</td>
  <td class=xl74 >1</td>
  <td class=xl74>50, 128, 800, 800</td>
  <td class=xl75 >1573.45</td>
  <td class=xl76 >2.44</td>
  <td class=xl75 >1576.23</td>
  <td class=xl76 >2.51</td>
  <td class=xl94 >1.00</td>
  <td class=xl95 >0.97</td>
 </tr>
 <tr>
  <td height=20 class=xl77 style='height:15.0pt;border-top:none'>rnn_tanh</td>
  <td class=xl67>1</td>
  <td class=xl67>2</td>
  <td class=xl6>50, 128, 800, 800</td>
  <td class=xl68>836.43</td>
  <td class=xl69>4.63</td>
  <td class=xl68>830.33</td>
  <td class=xl69>4.67</td>
  <td class=xl96>0.99</td>
  <td class=xl95 >0.99</td>
 </tr>
 <tr>
  <td height=20 class=xl77 style='height:15.0pt;border-top:none'>rnn_tanh</td>
  <td class=xl67>5</td>
  <td class=xl67>1</td>
  <td class=xl6>50, 128, 800, 800</td>
  <td class=xl68>381.32</td>
  <td class=xl69>11.44</td>
  <td class=xl68>379.88</td>
  <td class=xl69>11.50</td>
  <td class=xl96>1.00</td>
  <td class=xl95 >1.00</td>
 </tr>
 <tr>
  <td height=20 class=xl77 style='height:15.0pt;border-top:none'>rnn_tanh</td>
  <td class=xl67>5</td>
  <td class=xl67>2</td>
  <td class=xl6>50, 128, 800, 800</td>
  <td class=xl68>159.76</td>
  <td class=xl69>24.92</td>
  <td class=xl68>149.86</td>
  <td class=xl69>24.90</td>
  <td class=xl96>0.94</td>
  <td class=xl95 >1.00</td>
 </tr>
 <tr>
  <td height=20 class=xl73 style='height:15.0pt'>rnn_relu</td>
  <td class=xl74 >1</td>
  <td class=xl74 >1</td>
  <td class=xl74>50, 128, 800, 800</td>
  <td class=xl75 >1536.55</td>
  <td class=xl76 >2.65</td>
  <td class=xl75 >1540.29</td>
  <td class=xl76 >2.75</td>
  <td class=xl94 >1.00</td>
  <td class=xl95 >0.96</td>
 </tr>
 <tr>
  <td height=20 class=xl77 style='height:15.0pt;border-top:none'>rnn_relu</td>
  <td class=xl67>1</td>
  <td class=xl67>2</td>
  <td class=xl6>50, 128, 800, 800</td>
  <td class=xl68>805.00</td>
  <td class=xl69>5.09</td>
  <td class=xl68>807.68</td>
  <td class=xl69>5.06</td>
  <td class=xl96>1.00</td>
  <td class=xl95 >1.01</td>
 </tr>
 <tr>
  <td height=20 class=xl77 style='height:15.0pt;border-top:none'>rnn_relu</td>
  <td class=xl67>5</td>
  <td class=xl67>1</td>
  <td class=xl6>50, 128, 800, 800</td>
  <td class=xl68>373.27</td>
  <td class=xl69>12.41</td>
  <td class=xl68>377.79</td>
  <td class=xl69>12.32</td>
  <td class=xl96>1.01</td>
  <td class=xl95 >1.01</td>
 </tr>
 <tr height=19 style='height:14.4pt'>
  <td height=19 class=xl77 style='height:14.4pt;border-top:none'>rnn_relu</td>
  <td class=xl67>5</td>
  <td class=xl67>2</td>
  <td class=xl6>50, 128, 800, 800</td>
  <td class=xl68>154.21</td>
  <td class=xl69>26.93</td>
  <td class=xl68>153.80</td>
  <td class=xl69>26.61</td>
  <td class=xl96>1.00</td>
  <td class=xl95 >1.01</td>
 </tr>
</table>

**Comment by ciyongch** [comment]
@zixuanweeei Thanks for fixing the perf drop of LSTM. It's ok to remove "WIP" from title now.

**Comment by zixuanweeei** [comment]
@TaoLv Please take some reviews on this PR. Thanks.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 2]
```diff
@@ -73,64 +93,69 @@ static void ConcatData(mkldnn::memory::format src_format,
...
                        int concat_dimension,
-                       std::vector<void*> srcs_data,
-                       const mkldnn::memory &dst) {
+                       const std::vector<void*> &srcs_data,
+                       const mkldnn::memory &dst,
+                       std::vector<mkldnn::memory> *tmp_src_mems) {
   auto cpu_engine = CpuEngine::Get()->get_engine();
   std::vector<mkldnn::memory::primitive_desc> srcs_pd;
-  std::vector<mkldnn::memory> srcs;
+  bool initialized = tmp_src_mems->size() > 0;
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: const bool initialized?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 3]
```diff
@@ -73,64 +93,69 @@ static void ConcatData(mkldnn::memory::format src_format,
...
-      break;
-    default:
-      LOG(FATAL) << "unknown RNN mode " << mode;
-      break;
-  }
+  int n_gates = 0, n_states = 0;
+  GetMKLDNNRNNAlgo(mode, &n_gates, &n_states);
+  int n_bias = mode == rnn_enum::kGru ? n_gates + 1 : n_gates;
+  // sizes of single gates from a single cell
+  const size_t weights_size_0 = direction * (input_size + hidden_size) * hidden_size;
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: assign int to size_t?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 4]
```diff
@@ -225,55 +241,66 @@ static void MKLDNNRNNForwardSingleLayerBi(bool state_outputs,
...
-      AdjustGruBiasGateOrder(bx, H);
-      AdjustGruBiasGateOrder(back_bx, H);
-      AdjustGruBiasGateOrder(bh, H);
-      AdjustGruBiasGateOrder(back_bh, H);
+      has_adjusted = true;
     }
-    auto src_wx = (*concat_weight_memory)[2 * layer_index];
-    auto src_wh = (*concat_weight_memory)[2 * layer_index + 1];
+    auto src_wx = mkldnn_mems->concat_weight_memory[2 * layer_index];
+    auto src_wh = mkldnn_mems->concat_weight_memory[2 * layer_index + 1];
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: reference or copy?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 5]
```diff
@@ -225,55 +241,66 @@ static void MKLDNNRNNForwardSingleLayerBi(bool state_outputs,
...
+      const index_t single_b_sz = nbias * H;
+      #pragma omp parallel for num_threads(omp_threads)
+      for (int j = 0; j < H; j++) {
+        user_bias[j + H] = bx[j] + bh[j];
+        user_bias[single_b_sz + j + H] = back_bx[j] + back_bh[j];
+        user_bias[j] = bx[j + H] + bh[j + H];
+        user_bias[single_b_sz + j] = back_bx[j + H] + back_bh[j + H];
+      }
+      #pragma omp parallel for num_threads(omp_threads)
+      for (int j = 2 * H; j < 3 * H; j++) {
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Is it possible to merge this for loop to the above one? I can see they have the same steps but not sure if there is any dependency.

Comment (zixuanweeei) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Yep, we can merge these two into one loop. Both variants have the same performance. They cost about ~18 us with `hidden_size=4096` on 1 socket of SkyLake 8180 .

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 3 above)
Comment (zixuanweeei) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: The input params of `GetMKLDNNRNNCacheMemorySize` are set to be `const size_t` type. The intermdiate results of `int` type may overflow.

**Comment by zixuanweeei** [comment]
Sorry for the inconvenience. This PR has been moved to #15741.

**Pull request closed without merging**